feat(nydusify): add zero-disk streaming image transfer

daiyongxuan opened the pull request:
Implement a new streaming copy mechanism for image transfer that enables direct source-to-target streaming without using intermediate disk storage.

Key features:

Add StreamCopy for concurrent pull/push operations Introduce streamTransferManager for orchestrating streaming workflows Support chunked transfer with progress tracking
Add HTTP/HTTPS fallback with automatic retries
Support concurrent layer processing with configurable concurrency Retain compatibility with traditional transfer mode Performance improvements:

Eliminate need for intermediate disk storage
Lower memory footprint via streaming buffers
Enable concurrent image layer processing
Add progress logging for large transfers
Implementation details:

Introduce enableStreamTransfer configuration option Add httpModeManager for connection switching logic Optimize for both small and large file transfers
Add robust error handling and retry logic
Integrate with existing provider/resolver framework This significantly improves performance for large image transfers by avoiding disk I/O bottlenecks.

## Relevant Issue (if applicable)
_If there are Issues related to this PullRequest, please list it._

## Details
_Please describe the details of PullRequest._

## Types of changes

_What types of changes does your PullRequest introduce? Put an `x` in all the boxes that apply:_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Go over all the following points, and put an `x` in all the boxes that apply._

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

daiyongxuan committed:
feat(nydusify): add zero-disk streaming image transfer

Implement a new streaming copy mechanism for image transfer that enables direct source-to-target streaming without using intermediate disk storage.

Key features:
- Add `StreamCopy` for concurrent pull/push operations
- Introduce `streamTransferManager` for orchestrating streaming workflows
- Support chunked transfer with progress tracking
- Add HTTP/HTTPS fallback with automatic retries
- Support concurrent layer processing with configurable concurrency
- Retain compatibility with traditional transfer mode

Performance improvements:
- Eliminate need for intermediate disk storage
- Lower memory footprint via streaming buffers
- Enable concurrent image layer processing
- Add progress logging for large transfers

Implementation details:
- Introduce `enableStreamTransfer` configuration option
- Add `httpModeManager` for connection switching logic
- Optimize for both small and large file transfers
- Add robust error handling and retry logic
- Integrate with existing provider/resolver framework

This significantly improves performance for large image transfers by avoiding disk I/O bottlenecks.

daiyongxuan committed:
nydusify: fix golangci-lint issues

daiyongxuan committed:
nydusify: fix golangci-lint issues

daiyongxuan committed:
debug: revert to disable stream copy version

daiyongxuan committed:
feat(nydusify): add zero-disk streaming image transfer
Implement a new streaming copy mechanism for image transfer that enables direct source-to-target streaming without using intermediate disk storage.

Key features:

Add StreamCopy for concurrent pull/push operations
Introduce streamTransferManager for orchestrating streaming workflows
Support chunked transfer with progress tracking
Add HTTP/HTTPS fallback with automatic retries
Support concurrent layer processing with configurable concurrency
Retain compatibility with traditional transfer mode
Performance improvements:

Eliminate need for intermediate disk storage
Lower memory footprint via streaming buffers
Enable concurrent image layer processing
Add progress logging for large transfers
Implementation details:

Introduce enableStreamTransfer configuration option
Add httpModeManager for connection switching logic
Optimize for both small and large file transfers
Add robust error handling and retry logic
Integrate with existing provider/resolver framework
This significantly improves performance for large image transfers by avoiding disk I/O bottlenecks.

NehemiahMi commented on contrib/nydusify/pkg/converter/provider/provider.go:
```diff
@@ -231,3 +233,107 @@ func (pvd *Provider) NewRemoteCache(ctx context.Context, ref string) (context.Co
+func (pvd *Provider) FetchImageInfo(ctx context.Context, ref string) error {
+	resolver, err := pvd.Resolver(ref)
+	if err != nil {
+		return err
```
这里最好也使用 errors.Wrap ,使得后面好追踪

NehemiahMi commented on contrib/nydusify/pkg/copier/copier.go:
```diff
@@ -67,6 +78,340 @@ type output struct {
+			m.provider.UsePlainHTTP()
+
+			if err := m.refreshPusher(ctx); err != nil {
+				return err
```
errors.warp

daiyongxuan committed:
refactor: wrap errors with errors.Wrap and extract magic values to constants

NehemiahMi commented on contrib/nydusify/pkg/copier/copier.go:
```diff
@@ -67,6 +78,340 @@ type output struct {
+	return nil
+}
+
+func enableStreamCopy(opt Opt) bool {
```
PushChunkSize 原来就有意义，感觉还是用新的 参数来确定是否使用 streaming image transfer 而不是直接用复用这个值比较好。比如用户想使用 chunk push 而不是用 streaming image transfer 就只能用默认值，这是不对的

daiyongxuan committed:
feat: add a cli flag

daiyongxuan committed:
smoke: Add a smoke test for stream copy

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -1160,6 +1160,11 @@ func main() {
 					Usage: "Chunk size for pushing a blob layer in chunked",
 				},
+				&cli.BoolFlag{
+					Name:  "enable-stream-copy",
```
The flag is unnecessary, it should be a generic optimization.

imeoer commented on contrib/nydusify/pkg/converter/provider/provider.go:
```diff
@@ -34,6 +36,14 @@ import (
+	// Docker manifest list media type (legacy)
+	DockerManifestListMediaType = "application/vnd.docker.distribution.manifest.list.v2+json"
+	// Docker manifest media type (legacy)
+	DockerManifestMediaType = "application/vnd.docker.distribution.manifest.v2+json"
```
These two media type has been defined in `github.com/containerd/containerd/images`.

daiyongxuan committed:
Refractor: Create stream copy Implementation

daiyongxuan committed:
fix: resolve syntax errors after rebase

- Remove duplicate function definitions
- Fix StreamCopy function signature and calls
- Restore httpModeManager for stream transfer
- Update import statements for consistency

daiyongxuan committed:
style: fix whitespace formatting

daiyongxuan committed:
fix: resolve compatibility issues

imeoer commented:
Dup with https://github.com/dragonflyoss/nydus/pull/1765
